### PR TITLE
Week02 submissions

### PR DESCRIPTION
**Pull request by AhsanShafiqShawon**
Please review my code

### CODE REVIEW

File: week02/MinimumWindowSubstring.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        int L = 0, R = 0, minLen = INT_MAX;
+        int startingIndex = 0;
+        unordered_map<char, int> mp;
+
+        for(auto c : t) mp[c]++;
+        int cntOfExpectedChar = mp.size();
+
+        while(R < s.size()) {
+            mp[s[R]]--;
+            cntOfExpectedChar -= (mp[s[R]] == 0);
```
Comment (Sadman007) [inline week02/MinimumWindowSubstring.cpp:13]:
```suggestion
            if (mp[s[R]] == 0) {
                cntOfExpectedChar--;
            }
```

This one is more readable in this case.

File: week02/ValidParentheses.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                if(stck.empty() or (stck.top() == '(' and s[i] != ')')
+                or (stck.top() == '{' and s[i] != '}') or
+                (stck.top() == '[' and s[i] != ']'))
+                    return false;
+                stck.pop();
+            }
+        }
+        return stck.empty();
+    }
+};
```
Comment (Sadman007) [inline week02/ValidParentheses.cpp:18]:
```c++
// TC: O(N)
// MC: O(1)

class Solution {
public:
    bool isValid(string s) {
      unordered_map<char, char> close_to_open({{'}', '{'}, {')', '('}, {']', '['}});
      
      stack <char> open;
      for (auto &ch : s) {
        if (ch == '(' or ch == '{' or ch == '[') {
            open.push(ch);
        }
        else {
            if (!open.empty() && open.top() == close_to_open[ch]) open.pop();
            else return false;
        }
      }
        
      return open.empty();
    }
};
```

I would recommend this solution. The main difference between these two solutions is: mine is extendable. If in the future a new pair of parenthesis arrives, I can just add it in the map and it's done.

Comment (AhsanShafiqShawon) [reply week02/ValidParentheses.cpp:18]: Thanks for the recommendation.

File: week02/MinimumWindowSubstring.cpp (hunk 1 above)
Comment (AhsanShafiqShawon) [reply week02/MinimumWindowSubstring.cpp:13]: Thanks for the recommendation.

**Pull request merged**